main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -82,21 +88,109 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
   }
 
   /**
-    * Performs type inference and reassembly on all definitions in the given AST root.
+    * Performs type inference and reassembly on all instances in the given AST root.
     *
     * Returns [[Err]] if a definition fails to type check.
     */
-  private def visitDefs(root: ResolvedAst.Root)(implicit flix: Flix): Validation[Map[Symbol.DefnSym, TypedAst.Def], TypeError] = {
+  private def visitInstances(root: ResolvedAst.Root, classes: Map[Symbol.ClassSym, TypedAst.Class])(implicit flix: Flix): Validation[MultiMap[Symbol.ClassSym, TypedAst.Instance], TypeError] = {
+
+    /**
+      * Reassembles a single instance.
+      */
+    def visitInstance(inst: ResolvedAst.Instance): Validation[TypedAst.Instance, TypeError] = inst match {
+      case ResolvedAst.Instance(doc, mod, sym, tpe, tconstrs, defs0, loc) =>
+        for {
+          defs <- Validation.traverse(defs0)(visitDefn(_, root))
+        } yield TypedAst.Instance(doc, mod, sym, tpe, tconstrs, defs, loc)
+    }
+
     /**
-      * Performs type inference and reassembly on the given definition `defn`.
+      * Checks for overlap of instance types, assuming the instances are of the same class.
       */
-    def visitDefn(defn: ResolvedAst.Def): Validation[TypedAst.Def, TypeError] =
-      typeCheckDef(defn, root) map {
-        case (defn, subst) => defn
+    def checkOverlap(inst1: TypedAst.Instance, inst2: TypedAst.Instance)(implicit flix: Flix): Validation[Unit, TypeError] = {
+      if (Unification.unifyTypes(inst1.tpe, inst2.tpe).isOk) {
```

Comment:
Is this what the note does? I would have thought it would use the leq?

Reply:
We don't use `leq` here since we want a symmetric comparison, and because we disregard constraints.

This means that:
1. If `C[Int]` and `C[a]` are both declared, there will be an error, regardless of which is declared first.
1. If `C[List[a]]` and `C[List[a]] with [a : C]` are both declared, there will be an error; there can only be at most one implementation for any given type, regardless of which other classes have implementations for it.